### PR TITLE
crimson/osd: support "dump_pgstate_history" command and update legacy conf vars

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
see also 076c4d752c62276e78129c9a82aa50dfca83c0f0

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-08-28_06:18:09-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-08-28_13:09:24-rados-wip-kefu-testing-2020-08-28-1247-distro-basic-smithi/

File: src/common/config.h [hunk 1]
```diff
@@ -124,6 +124,7 @@ struct md_config_t {
   int parse_buffer(ConfigValues& values, const ConfigTracker& tracker,
 		   const char* buf, size_t len,
 		   std::ostream *warnings);
+  void update_legacy_vals(ConfigValues& values);
```
Comment (rzarzynski) [inline src/common/config.h:127]: nit: a typo in the commit msg (`popolated`).

File: src/crimson/common/config_proxy.cc [hunk 2]
```diff
@@ -76,7 +76,11 @@ seastar::future<> ConfigProxy::parse_config_files(const std::string& conf_files)
       }
       return crimson::read_file(*path++).then([this](auto&& buf) {
         return do_change([buf=std::move(buf), this](ConfigValues& values) {
-          if (get_config().parse_buffer(values, obs_mgr, buf.get(), buf.size(), &std::cerr)) {
+          if (get_config().parse_buffer(values, obs_mgr,
```
Comment (rzarzynski) [inline src/crimson/common/config_proxy.cc:69]: same typo also in the second commit.

File: src/crimson/common/config_proxy.cc [hunk 3]
```diff
@@ -3,20 +3,10 @@
 
 #include "config_proxy.h"
 
-#if __has_include(<filesystem>)
```
Comment (rzarzynski) [inline src/crimson/common/config_proxy.cc:6]: :+1:

File: src/crimson/admin/admin_socket.cc [hunk 4]
```diff
@@ -397,8 +397,10 @@ class InjectArgsHook final : public AdminSocketHook {
 				      ceph::bufferlist&&) const final
   {
     std::vector<std::string> argv;
-    [[maybe_unused]] bool found = cmd_getval(cmdmap, "injected_args", argv);
-    assert(found);
+    bool found = cmd_getval(cmdmap, "injected_args", argv);
+    if (!found) {
```
Comment (rzarzynski) [inline src/crimson/admin/admin_socket.cc:401]: nit: do we need this variable?

File: src/crimson/admin/osd_admin.cc [hunk 5]
```diff
@@ -105,6 +105,30 @@ class FlushPgStatsHook : public AdminSocketHook {
...
+    AdminSocketHook{"dump_pgstate_history",
+                    "",
+                    "dump history of PGs' peering state"},
+    osd{osd}
+  {}
+  seastar::future<tell_result_t> call(const cmdmap_t&,
+                                      std::string_view format,
+                                      ceph::bufferlist&& input) const final
+  {
+    unique_ptr<Formatter> f{Formatter::create(format, "json-pretty", "json-pretty")};
```
Comment (rzarzynski) [inline src/crimson/admin/osd_admin.cc:121]: Is there a particular reason for avoiding `std::`? It's used below (`template std::unique_ptr ...`).

File: src/crimson/osd/osd.cc [hunk 6]
```diff
@@ -427,7 +427,7 @@ seastar::future<> OSD::start_asok_admin()
   return asok->start(asok_path).then([this] {
     return seastar::when_all_succeed(
       asok->register_admin_commands(),
-      asok->register_command(make_asok_hook<OsdStatusHook>(*this)),
+      asok->register_command(make_asok_hook<OsdStatusHook>(std::as_const(*this))),
```
Comment (rzarzynski) [inline src/crimson/osd/osd.cc:430]: :+1:

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/admin/osd_admin.cc (hunk 5 above)
Comment (tchaikov) [reply src/crimson/admin/osd_admin.cc:121]: copy-n-paste error. fixed.

File: src/crimson/admin/admin_socket.cc (hunk 4 above)
Comment (tchaikov) [reply src/crimson/admin/admin_socket.cc:401]: no. removed.

File: src/crimson/common/config_proxy.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/common/config_proxy.cc:69]: fixed.

File: src/common/config.h (hunk 1 above)
Comment (tchaikov) [reply src/common/config.h:127]: fixed.

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Pull request merged**